### DISCUSSION TITLE
Elasticsearch Ingest parser not overwritten

### DISCUSSION THREAD

**Discussion by stijnpieters** [Security-Onion-Solutions/securityonion#2526] [Unsupported Versions]
I am using SO2.3.21
so-managersearch node + 2 forward nodes

I am writing ingest parsers to ingest Sophos UTM logs. I kept getting errors saying that no match was found for a pattern I used to test how these parsers work. Using curl and the Elasticsearch API I found out the old ingest parser was still there, not the new one I had written (after many so-elastalert-restarts). 
I then manually deleted the existing ingest parser using curl -X DELETE so-manager:9200/_ingest/pipeline/my_pipeline_id and reran so-elasticsearch-restart.

2 Questions
1. Am I pushing my ingest parsers to elasticsearch correctly by letting your script handle it by running so-elasticsearch-restart?
2. Is this intended behaviour? Or is this something that should be fixed?

**Comment by senatorhotchkiss** [comment]
A general guide on doing the elastic parsing is also a very big request from my side. In v16 Logstash was the thing and very easy to add new custom input.

The documentation on ES parsing is a bit confusing. I still dont fully understand how the ingest pipelines, are called upon. They seem to refer to one another, but I still have not broken the code. 

A few examples of adding new pipelines, would be worth diamonds in my case.

**Reply by stijnpieters** [reply]
The pfsense one in the security onion install helped me out a lot, I just took a raw log and compared the fields from that to the 'filterlog' pipeline. 
I agree the docs are kind of confusing, but once you're confused enough yourself, you start to understand them :p

**Reply by senatorhotchkiss** [reply]
One of the main things, I can't seem to understand, is how incoming data is matched to the correct pipeline file.

Syslog--> goes to which ingest files and in what order?
Winlogbeat/filebeat-->Logstash(not doing any parsing)-->goes to which ingest files and in what order.

After writing little grok in the old SO w. Logstash, I feel stupid not being able to even understand the flow in SO2 😅

**Reply by stijnpieters** [reply]
what pipeline it goes to is usually defined by a "pipeline" operator, where you put an "if" in for example to match a source application or sth. When the source application is filterlog for example it goes to the "filterlog" pipeline. 

I constantly feel like I figured it out, and then I just run headfirst into a wall trying to build a decent pipeline

**Reply by senatorhotchkiss** [reply]
I found logstash files refering the ES parsers. Now it makes sense. Example: beats.common is set as output from logstash.

**Reply by stijnpieters** [reply]
And all syslog gets the "syslog" ingest parser as output. In /opt/so/saltstack/default/salt/elasticsearch/files/ingest/syslog you can see a "dissect", which just puts the message in a field. Then removes 2 fields and THEN applies a good old fashioned grok filter, which matches common syslog patterns. In these patterns it matches for example "source.application". For syslogs sent from a PfSense, this "source.application" is "filterlog", which is then matched in a "pipeline" operator a little lower in the file. Then it goes into the other pipeline and does the things parsers do. 

At least that's how I think it works now. 
I already managed to get into my own "sophos_utm" pipeline. Now I just need to manage to do some parsing.

**Reply by senatorhotchkiss** [reply]
In the logstash pipelines I found a few files:

cat 9034_output_syslog.conf


output {
  if [module] =~ "syslog" {
    elasticsearch {
      pipeline => "%{module}"
      hosts => "10.100.40.230"
      index => "so-syslog"
      template_name => "so-syslog"
      template => "/templates/so-syslog-template.json"
      template_overwrite => true
    }
  }
}

To me it looks like they are passing stuff on to the ES and defining the pipeline. It is for sure very complex :)

**Reply by weslambert** [reply]
The flow, in general when running Logstash goes like:

Filebeat -> Logstash Beats input -> Logstash Redis output -> Logstash Redis input -> Logstash output to ES Ingest Pipeline base on module/dataset -> ES Ingest Pipeline/parsing (potentially going to other pipelines, based on event type/details) -> common pipeline -> ES index defined in Logstash output/Filebeat input, or altered depending on event type.

**Reply by senatorhotchkiss** [reply]
I got it figured now. The logstash selection was the missing link for me. 
If I have some input for improvements, how should I proceed with sharing this?

**Reply by weslambert** [reply]
You are welcome to submit a PR for RTD, or elsewhere.  Otherwise, you can share your feedback here.

**Reply by senatorhotchkiss** [reply]
Allow GeoIP of clients to a Remote Desktop Gateway, by adding this line:

win.eventlogs
{ "rename":      { "field": "winlog.user_data.IpAddress",                                     "target_field": "source.ip",                "ignore_missing": true  } }

I also wrote logstash for parsing Exchange/IIS logfiles and RRAS. The output of these are of very high value to SOC.

**Comment by weslambert** [comment]
I see if I can put a practical example (post, etc) together soon.

On Mon, Jan 4, 2021 at 11:34 AM senatorhotchkiss <notifications@github.com>
wrote:

> A general guide on doing the elastic parsing is also a very big request
> from my side. In v16 Logstash was the thing and very easy to add new custom
> input.
>
> The documentation on ES parsing is a bit confusing. I still dont fully
> understand how the ingest pipelines, are called upon. They seem to refer to
> one another, but I still have not broken the code.
>
> A few examples of adding new pipelines, would be worth diamonds in my case.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Security-Onion-Solutions/securityonion/discussions/2526#discussioncomment-260464>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEAM3KA3SKY2LUWLZNM3OKDSYHUY5ANCNFSM4VTHCCFQ>
> .
>


-- 
https://twitter.com/therealwlambert
https://securityonion.net/
